### PR TITLE
Scheduled banners

### PR DESCRIPTION
**Pull request by elarb**
- Makes it possible to schedule banners upfront given a date range from the admin panel
- Makes use of the newly added `/web/banners` end-point, that corresponds to [this ](https://github.com/AreaFiftyLAN/lancie-api/pull/450) PR on the API
- A new `text` tab on the admin panel has been added, which includes the newly added banner scheduling panel
- The `text` tab should contain other panels that schedule/change data on the home page (or eventually other pages) of the website, that we might want to add in the feature (https://github.com/AreaFiftyLAN/lancie-api/issues/280)
- Along the way, I separated icons used in a custom `lancie-admin` icon set, to improve performance 

Make sure to check out to [this](https://github.com/AreaFiftyLAN/lancie-api/tree/feature-scheduled_banner) branch when testing the newly added functionality.

Known issue: `update` buttons are shown everywhere after a delete. Reason for this is that a delete causes an `lancie-ajax` request and this triggers the observer for each element and thus assumes that it has been updated.

Closes https://github.com/AreaFiftyLAN/lancie-api/issues/268

### CODE REVIEW

File: src/lancie-admin-banner/lancie-admin-banner-scheduler.html [hunk 1]
```diff
@@ -0,0 +1,150 @@
+<link rel="import" href="../../bower_components/polymer/polymer.html">
+<link rel="import" href="../../bower_components/lancie-ajax/lancie-ajax.html">
+<link rel="import" href="../../bower_components/paper-input/paper-textarea.html">
+<link rel="import" href="../../bower_components/paper-icon-button/paper-icon-button.html">
+<link rel="import" href="../../bower_components/paper-input/paper-input.html">
+
+<!--Element that makes it possible to schedule banners by providing start- and enddates and a corresponding text for
+the banner within that timeframe-->
```
Comment (TimvdLippe) [inline src/lancie-admin-banner/lancie-admin-banner-scheduler.html]: Documentation 👏 👏 👏

File: src/lancie-admin-content.html [hunk 2]
```diff
@@ -43,11 +44,12 @@
...
       <lancie-admin-types data-route="types" data-path="lancie-admin-endpoint/lancie-admin-pages/lancie-admin-types.html"></lancie-admin-types>
       <lancie-admin-seats data-route="seats" data-path="lancie-admin-endpoint/lancie-admin-pages/lancie-admin-seats.html"></lancie-admin-seats>
+      <lancie-admin-hptext data-route="text" data-path="lancie-admin-endpoint/lancie-admin-pages/lancie-admin-hptext.html"></lancie-admin-hptext>
       <lancie-admin-rfid-assign data-route="rfidassign" data-path="lancie-admin-rfid/lancie-admin-rfid-assign.html"></lancie-admin-rfid-assign>
       <lancie-admin-rfid-alcohol data-route="alcoholcheck" data-path="lancie-admin-rfid/lancie-admin-rfid-alcohol.html"></lancie-admin-rfid-alcohol>
       <lancie-admin-rfid-consumption data-route="consumptions" data-path="lancie-admin-rfid/lancie-admin-rfid-consumption.html"></lancie-admin-rfid-consumption>
 
-      <lancie-admin-404 data-route="404" data-path="lancie-admin-404.html"></lancie-404>
+      <lancie-admin-404 data-route="404" data-path="lancie-admin-404.html"></lancie-admin-404>
```
Comment (TimvdLippe) [inline src/lancie-admin-content.html:52]: 🤔

File: src/lancie-admin-endpoint/lancie-admin-pages/lancie-admin-hptext.html [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+      }
+
+      paper-fab {
+        position: absolute;
+        right: 24px;
+        bottom: 24px;
+        --paper-fab-background: var(--lancie-button-color);
+      }
+
+      /*TODO: Should be in global scope*/
```
Comment (TimvdLippe) [inline src/lancie-admin-endpoint/lancie-admin-pages/lancie-admin-hptext.html]: Hm, that's tough. Have to think about how to do that

File: src/lancie-admin-endpoint/lancie-admin-pages/lancie-admin-hptext.html [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+        this.push('banners', {startDate: latest, endDate: latest});
+      }
+
+      refreshBanners() {
+        this.$.bannersAjax.generateRequest();
+        this.sortOnStartDate();
+      }
+
+      sortOnStartDate() {
+        this.banners.sort((x, y) => x.startDate - y.startDate);
```
Comment (TimvdLippe) [inline src/lancie-admin-endpoint/lancie-admin-pages/lancie-admin-hptext.html]: Does this trigger the bindings update? Else it has to be `this.banners = this.banners.sort`

File: src/lancie-admin-endpoint/lancie-admin-pages/lancie-admin-hptext.html [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+      }
+
+      getLatestDate() {
+        if (this.banners.length > 0) {
+          this.banners.sort((x, y) => x.startDate - y.startDate);
```
Comment (TimvdLippe) [inline src/lancie-admin-endpoint/lancie-admin-pages/lancie-admin-hptext.html]: Why are we sorting again? And if so, it's better to call `sortOnStartDate` here

File: src/lancie-admin-icons.html [hunk 6]
```diff
@@ -0,0 +1,22 @@
+<link rel="import" href="../bower_components/iron-icon/iron-icon.html">
```
Comment (TimvdLippe) [inline src/lancie-admin-icons.html:1]: I like it!

**Review by TimvdLippe — COMMENTED** [review]
Statically reviewed. Seems like a nice contribution!

**Review by DieuwerH — COMMENTED** [review]
Error messages for the date fields are not visible / cut off

Would suggest putting the add banner button next to the remove all button

The remove all button has more space to the left than the rest of the elements

The dates in the the timeline only update when refreshing the page, not when changing the values

The flow of changing an item does not feel natural, you would expect to press the pencil to be able to change a value, not to save them. So by default all fields are disabled until the pen button is pressed. The pen image should change to a save image and upon pressing it the item is saved; disabling all the fields again and changing the button's image back to a pencil.

**Comment by elarb** [comment]
@DieuwerH 

- Fixed, thanks
- Matter of preference
- I noticed as well, changed (it was `1px` btw)
- Could you show an example?
- Agreed, something I also was not sure about. Changed to an `update` button now.

**Comment by martijnjanssen** [comment]
Can't you just fire the ajax request again to refresh the items when one is added or removed? Or at least a button to refresh the timeline display?

**Comment by martijnjanssen** [comment]
I also recommend not placing the `lancie-admin-hptext.html` in the `lancie-admin-pages` folder, the name is not very clear for that folder, but it is more aimed at the pages with a datatable. It would be better to just make a folder for `lancie-admin-homepage-text` and put everything in there.

**Comment by elarb** [comment]
@martijnjanssen Agreeing with the naming, will change that, but I don't really see how it doesn't fit in that folder.

File: src/lancie-admin-content.html (hunk 2 above)
Comment (martijnjanssen) [reply src/lancie-admin-content.html:52]: Yeah that one has been fixed in 4 non-merged branches now 🤔

**Comment by martijnjanssen** [comment]
- The `paper-dialog` is behind the drawer, this is easily fixed by using our own `lancie-dialog`.
- The circular add-button is very nice, but it doesn't really fit with the rest of our pages. Why not just use a button next to the `remove all`-button?
- I still thing we have to do something about the layout of the fields to make it more mobile or scaling friendly.
- And lastly: Nice!

**Comment by elarb** [comment]
`lancie-dialog` seems to behave the same: 

chrome_2017-12-01_21-08-54

I fixed it though by appending the dialog to the document body after the element is being added:

chrome_2017-12-01_21-19-47

**Comment by elarb** [comment]
This is how `lancie-dialog` looks like when appending it to the document body:

chrome_2017-12-02_18-04-12

For now, we can either use `lancie-dialog` and live with the fact that it is behind the drawer, or we can keep it how it is (use `paper-dialog`) and fix this in a later PR (e.g. when we update `lancie-dialog`)?

**Comment by elarb** [comment]
Alright, I think this is ready to be merged now

**Review by MatthijsKok — APPROVED** [review]
Love it!
